fuse_loss/test/test_qwt_loss_plot.cpp

```diff
@@ -0,0 +1,206 @@
+/*
+ * Software License Agreement (BSD License)
+ *
+ *  Copyright (c) 2019, Clearpath Robotics
+ *  All rights reserved.
+ *
+ *  Redistribution and use in source and binary forms, with or without
+ *  modification, are permitted provided that the following conditions
+ *  are met:
+ *
+ *   * Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ *   * Redistributions in binary form must reproduce the above
+ *     copyright notice, this list of conditions and the following
+ *     disclaimer in the documentation and/or other materials provided
+ *     with the distribution.
+ *   * Neither the name of the copyright holder nor the names of its
+ *     contributors may be used to endorse or promote products derived
+ *     from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ *  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ *  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ *  FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ *  COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ *  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ *  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ *  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ *  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+#include <fuse_loss/arctan_loss.h>
+#include <fuse_loss/cauchy_loss.h>
+#include <fuse_loss/dcs_loss.h>
+#include <fuse_loss/fair_loss.h>
+#include <fuse_loss/geman_mcclure_loss.h>
+#include <fuse_loss/huber_loss.h>
+#include <fuse_loss/pseudo_huber_loss.h>
+#include <fuse_loss/softlone_loss.h>
+#include <fuse_loss/tolerant_loss.h>
+#include <fuse_loss/trivial_loss.h>
+#include <fuse_loss/tukey_loss.h>
+#include <fuse_loss/welsch_loss.h>
+
+#include <fuse_loss/qwt_loss_plot.h>
+
+#include <gtest/gtest.h>
+
+#include <QApplication>
+
+#include <vector>
+
+
+class QwtLossPlotTest : public testing::Test
+{
+public:
+  QwtLossPlotTest()
+  {
+    // Generate samples:
+    const double step{ 0.01 };
+    const size_t half_samples{ 1000 };
+    const double x_min = -(half_samples * step);
+
+    const size_t samples{ 2 * half_samples + 1 };
+
+    residuals.reserve(samples);
+
+    for (size_t i = 0; i < samples; ++i)
+    {
+      residuals.push_back(x_min + i * step);
+    }
+  }
+
+  std::vector<double> residuals;
+};
+
+TEST_F(QwtLossPlotTest, PlotLossQt)
```

Comment:
Given that there are no actual tests in this unit test, would you want to convert this to just a plain executable?

Reply:
One reason I decided to create a test instead of an executable is that I can use `<test_depend>`s, so I don't bring Qt and Qwt dependencies in for the regular package installation.

Another option could be to consolidate this test with the other test for loss functions. That way the test will check the `rho` values are correct for all the residuals evaluated to generate the plots, which is likely a better and more exhaustive test, and still quick. The only reason I didn't do that was because I expected some complexity in the logic that generates the plots, but I can try simplifying it as suggested in the other review comments. But there's still the problem with the headless machine failing to run this test, so that should be resolved before doing something like that.

Thoughts?

Reply:
The dependency concern is valid. There is always gnuplot as an option. I believe it will run in headless mode. And it looks like there is a rosdep rule for it already. https://github.com/ros/rosdistro/blob/master/rosdep/base.yaml#L1059

Or you can mark it as deprecated and whoever wants to run it will just need to modify that line. That works for me as well.

Reply:
Not sure about **gnuplot**, but with **GT/Qwt** I can have the interactive mode, that is useful to zoom and pan, so I can inspect the plots better if I enable the interactive mode with the `BUILD_WITH_INTERACTIVE_TESTS` option.

I've just added another option to enable the plotting tests, that default to `OFF`, so the checks should pass now: `BUILD_WITH_PLOT_TESTS`.

This way we can easily enable the test from the command line doing `cmake . -DBUILD_WITH_PLOT_TEST=ON` from the `build` directory.
Or `cmake . -DBUILD_WITH_PLOT_TESTS=ON -DBUILD_WITH_INTERACTIVE_TESTS=ON` if we also want to interact with the plots using the **Qwt** plots.